### PR TITLE
fix an issue that scheduling doesn't respect NodeLost status of a node

### LINKED ISSUE #67733: node wasn't tainted unreachable when TaintNodesByCondition is enabled
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When TaintNodesByCondition is enabled, if a node lost connection to apiserver, node lifecycle manager doesn't add "unreachable" taint to the node.
So that results in scheduler still scheduled pods onto that "lost" node, in a Pending status.

**What you expected to happen**:

If a node is lost, an "unreachable" taint should be added onto the node, so that new pods won't be scheduled onto the "unreachable" node.
And if the node gets recovered, the taint should be removed. 

**How to reproduce it (as minimally and precisely as possible)**:

- enable TaintNodesByCondition - it's default behavior of current master branch
- create a 1-node cluster
- create a deployment with 10 replicas, and they landed on the only node with running status
- SSH into the node and break connectivity to apiserver (`iptables --insert OUTPUT --destination <apiserver ip> --jump REJECT`)
- wait for the node-1 to be NotReady (not that at this moment, the 10 pods are still running status, but internally the Ready condition is false, I guess it's node lifecycle manager modifying them when it's not getting heartbeat from kubelet within a time period)
- then create node-2 and join the cluster
- a few mins later, 5 pods were created with running status onto node-2, and 5 pods still get scheduled to node-1 with Pending status. And old 10 pods switches to Unknown status.

If I disable TaintNodesByCondition and repeat the whole process, result of last step above is different:

- a few mins later, another 10 pods were created with running status onto node-2. And old 10 pods switches to Unknown status.

**Anything else we need to know?**:

#67536 might be another symptom due to the same bug.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.12 (or v1.1x where TaintNodesByCondition is enabled manually)
- Cloud provider or hardware configuration: kubeadm
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`):
- Install tools: kubeadm
- Others:

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What this PR does / why we need it**:

- if Node is in UnknowStatus, apply unreachable taint with NoSchedule effect
- some internal data structure refactoring
- update unit test

**Which issue(s) this PR fixes**:
Fixes #67733, and very likely #67536

**Special notes for your reviewer**:

See detailed reproducing steps in #67733.

**Release note**:
```release-note
Apply unreachable taint to a node when it lost network connection.
```

### CODE REVIEW

**Comment by k82cn** [comment]
/approve

LGTM, to see if other one has comments :)

**Comment by k82cn** [comment]
xref https://github.com/kubernetes/kubernetes/issues/67300

**Comment by Huang-Wei** [comment]
/hold

Let me take a look at the test failure.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (bsalamat) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: If you look at line 70, it seems that NodeUnreachable should be a NoExecute taint, i.e. causes eviction of running pods. If that's true, we should not always use NoSchedule here, at line 455.

Comment (liqlin2015) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: `NoSchedule` is used for scheduling, which is different. This fix can address issue that scheduler still schedule pods to NotReady node.
Another question, did you try shutting down node and see whether both `node.kubernetes.io/not-ready:NoSchedule` and `node.kubernetes.io/not-ready:NoExecute` taint are there?  Code https://github.com/kubernetes/kubernetes/blob/master/pkg/util/taints/taints.go#L277 seems did not do the job.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (ceshihao) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Does it mean NodeCondition with type = `NodeReady` and status = `ConditionUnknown` should be considered as `NodeUnreachable`?

It is a little ambiguous. Perhaps a new type `NodeReachable` is more reasonable.
https://github.com/kubernetes/kubernetes/blob/5a16163c87fe2a90916a51b52771a668bcaf2a0d/staging/src/k8s.io/api/core/v1/types.go#L4019-L4023

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (Huang-Wei) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: @bsalamat `UnreachableTaintTemplate` in line 70 (a.k.a unreachable/NoExecute taint) is applied when `TaintBasedEvictions` is enabled - although the name is ambiguous.

I think the idea here is to differentiate NoSchedule and NoExecute taint:
- TaintBasedEvictions controls `not-ready/NoExecute` and `unreachable/NoExecute`
- TaintNodesByCondition controls `xyz/NoSchedule`

And I found another bug that when TaintBasedEvictions and TaintNodesByCondition are both enabled, `xyz/NoSchedule` taint will **overwrite** existing `xyz/NoExecute` taints. It's because the logic below only check if the taintKey is found or not, regardless the effect:
https://github.com/kubernetes/kubernetes/blob/7a9a30a4f7eae900bffd4b477a431a999de350dd/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L474-L483

In latest commit, it's resolved.

So right now when TaintBasedEvictions and TaintNodesByCondition are both enabled, the behavior is like this:

- create a 2-node cluster
- create a deployment with 8 replicas - 4 landed on node1, 4 landed on node2
- break node1's network connectivity to apiserver
- node1 is applied with following taints:
    ```yaml
    spec:
      taints:
      - effect: NoSchedule
        key: node.kubernetes.io/unreachable
        timeAdded: 2018-08-23T19:21:58Z
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        timeAdded: 2018-08-23T19:22:03Z
    ```
- also pods landed on node1 are applied `NoExecute` toleration - means it will be evicted in 300 seconds:
    ```yaml
      tolerations:
      - effect: NoExecute
        key: node.kubernetes.io/not-ready
        operator: Exists
        tolerationSeconds: 300
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        operator: Exists
        tolerationSeconds: 300
    ```
- 300 seconds later, node lifecycle manager starts to evicts pods are spawn new pods - due to introduced `unreachable/NoSchedule` taint, new pods all landed to node2
    ```yaml
    # k get po -o wide
    NAME                     READY     STATUS        RESTARTS   AGE       IP            NODE          NOMINATED NODE
    pause-54dfd6bf9c-2nwrm   1/1       Running       0          8m9s      10.244.3.4    kube-node-2   <none>
    pause-54dfd6bf9c-6v62c   1/1       Running       0          44s       10.244.3.9    kube-node-2   <none>
    pause-54dfd6bf9c-8c7rh   1/1       Running       0          44s       10.244.3.8    kube-node-2   <none>
    pause-54dfd6bf9c-8nw4s   1/1       Terminating   0          8m9s      10.244.2.4    kube-node-1   <none>
    pause-54dfd6bf9c-b2sng   1/1       Terminating   0          8m9s      10.244.2.6    kube-node-1   <none>
    pause-54dfd6bf9c-bczdb   1/1       Running       0          8m9s      10.244.3.5    kube-node-2   <none>
    pause-54dfd6bf9c-fms72   1/1       Running       0          44s       10.244.3.7    kube-node-2   <none>
    pause-54dfd6bf9c-g7nnv   1/1       Running       0          8m9s      10.244.3.3    kube-node-2   <none>
    pause-54dfd6bf9c-gv8d5   1/1       Running       0          44s       10.244.3.10   kube-node-2   <none>
    pause-54dfd6bf9c-n5njz   1/1       Running       0          8m9s      10.244.3.6    kube-node-2   <none>
    pause-54dfd6bf9c-sfj6h   1/1       Terminating   0          8m9s      10.244.2.3    kube-node-1   <none>
    pause-54dfd6bf9c-zqtdc   1/1       Terminating   0          8m9s      10.244.2.5    kube-node-1   <none>
    # k get deploy
    NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
    pause     8         8         8            8           8m21s
    # k get rs
    NAME               DESIRED   CURRENT   READY     AGE
    pause-54dfd6bf9c   8         8         8         8m28s
    ```
- the 4 Terminating pods still shows is b/c kubelet in node1 can't talk to apiserver, hence still have those orphan containers living - but they doesn't count into replicas of the deployment. They are cleaned up gracefully after node1 gets recovered.

Comment (Huang-Wei) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: @liqlin2015 yeap, I see both taints. See above comment for more details.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (Huang-Wei) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: > Does it mean NodeCondition with type = NodeReady and status = ConditionUnknown should be considered as NodeUnreachable?

yeap.



> It is a little ambiguous. Perhaps a new type NodeReachable is more reasonable.

Agree, that's the ideal case. I would also like to see this kind of status :)

But that should be considered by sig/node, maybe as a new feature.

**Comment by Huang-Wei** [comment]
The test failure showed a corner case: the node is having condition `Ready` == `Unknown` and `Reason` == `NodeStatusNeverUpdated` - not quite sure how this can happen in real world: node is registered to the cluster, but afterwards it suddenly lose network hence could report even 1st heartbeat to apiserver?

```
0823 20:06:30.872] Recording: run_nodes_tests
I0823 20:06:30.872] Running command: run_nodes_tests
I0823 20:06:30.890] 
I0823 20:06:30.891] +++ Running case: test-cmd.run_nodes_tests 
I0823 20:06:30.893] +++ working dir: /go/src/k8s.io/kubernetes
I0823 20:06:30.895] +++ command: run_nodes_tests
I0823 20:06:30.903] +++ [0823 20:06:30] Testing kubectl(v1:nodes)
I0823 20:06:31.007] [32mcore.sh:1317: Successful get nodes {{range.items}}{{.metadata.name}}:{{end}}: 127.0.0.1:
I0823 20:06:31.162] (B[m[32mcore.sh:1319: Successful describe nodes 127.0.0.1:
I0823 20:06:31.162] Name:               127.0.0.1
I0823 20:06:31.162] Roles:              <none>
I0823 20:06:31.162] Labels:             <none>
I0823 20:06:31.162] Annotations:        node.alpha.kubernetes.io/ttl: 0
I0823 20:06:31.162] CreationTimestamp:  Thu, 23 Aug 2018 20:02:06 +0000
I0823 20:06:31.162] Taints:             node.kubernetes.io/unreachable:NoSchedule
I0823 20:06:31.163] Unschedulable:      false
I0823 20:06:31.163] Conditions:
I0823 20:06:31.163]   Type             Status    LastHeartbeatTime                 LastTransitionTime                Reason                   Message
I0823 20:06:31.163]   ----             ------    -----------------                 ------------------                ------                   -------
I0823 20:06:31.163]   Ready            Unknown   Thu, 23 Aug 2018 20:02:06 +0000   Thu, 23 Aug 2018 20:03:10 +0000   NodeStatusNeverUpdated   Kubelet never posted node status.
I0823 20:06:31.163]   OutOfDisk        Unknown   Thu, 23 Aug 2018 20:02:06 +0000   Thu, 23 Aug 2018 20:03:10 +0000   NodeStatusNeverUpdated   Kubelet never posted node status.
I0823 20:06:31.163]   MemoryPressure   Unknown   Thu, 23 Aug 2018 20:02:06 +0000   Thu, 23 Aug 2018 20:03:10 +0000   NodeStatusNeverUpdated   Kubelet never posted node status.
I0823 20:06:31.164]   DiskPressure     Unknown   Thu, 23 Aug 2018 20:02:06 +0000   Thu, 23 Aug 2018 20:03:10 +0000   NodeStatusNeverUpdated   Kubelet never posted node status.
```

Fix in this PR applied a unreachable/NoSchedule to this node, but latter testcase expect there is no taint.

Possible options are:
1. figure out why the testing node is in NodeStatusNeverUpdated status and give a fix
2. fix in the testcase that expect no taints
3. fix in ourside to exclude `Reason` == `NodeStatusNeverUpdated`

@bsalamat @k82cn any thought?

/sig testing

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: > I think the idea here is to differentiate NoSchedule and NoExecute taint:

Regarding `TaintBasedEviction`, node lifecycle already has feature to evict/kill pod when there's some issue to the node before toleration/taint; and then we leverage NoExecute to do that.

Regarding `TaintNodeByCondition`, there's a requirements to make pod tolerating some condition (e.g. network unavailable for who's using host network) which is hard code in scheduler before.


> taintutils.TaintSetFilter

+1

**Comment by Huang-Wei** [comment]
cc test/cmd owners 

@kubernetes/sig-cli-maintainers @kubernetes/sig-cli-pr-reviews

**Comment by Huang-Wei** [comment]
/cc @adohe @liggitt PTAL.

The story is that we decide to apply taint for nodes which are in Unknown status - see issue #67733 and #67536 for details if you want.

However, in test/cmd/xxx.sh, it's running based on a "fake" node (no kubelet, only API object Node was created), and then the node is marked with status == Unknown. So our change apply a taint to the "faked" node. It's all good but failed in `test/cmd/node-management.sh` where expect no taints for the node when doing some taint tests.

So I'd like you to review the changes I proposed in `test/cmd/node-management.sh`. Right now CI is all green. Thanks in advance.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (ravisantoshgudimetla) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: nit: s/candicates/candidates

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go
Comment (ravisantoshgudimetla) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: nit: s/unknow/unknown

Comment (Huang-Wei) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: done

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (Huang-Wei) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: done

File: test/cmd/node-management.sh
Comment (Huang-Wei) [reply test/cmd/node-management.sh]: done

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go
Comment (bsalamat) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Thanks, @Huang-Wei for the detailed explanation.

**Review by bsalamat — APPROVED** [review]
/lgtm

The change to node-management.sh looks good to me, but we should see what owners of that piece of code think.

**Comment by Huang-Wei** [comment]
according to changing history of node-management.sh, @liggitt might be the best person to review the change.

/assign @liggitt on changes of node-management.sh

File: test/cmd/node-management.sh
Comment (liggitt) [inline test/cmd/node-management.sh]: is it possible to make this just check for the `dedicated:PreferNoSchedule` taint? would like to avoid needing to adjust this test on every controller-managed taint change

File: test/cmd/node-management.sh
Comment (liggitt) [inline test/cmd/node-management.sh]: likewise, can we check for the absence of the `dedicated:PreferNoSchedule` taint, rather than explicitly listing all other present taints?

File: test/cmd/node-management.sh
Comment (Huang-Wei) [reply test/cmd/node-management.sh]: It's possible, but just current `kube::test::get_object_assert` implementation is fixed to a pattern to matching `^<expected string>$`. Let me try to work it around.

File: test/cmd/node-management.sh
Comment (Huang-Wei) [reply test/cmd/node-management.sh]: It's possible, but just current kube::test::get_object_assert implementation is fixed to a pattern to matching ^<expected string>$. Let me try to work it around.

Comment (Huang-Wei) [reply test/cmd/node-management.sh]: Resolved using go-template trick.

File: test/cmd/node-management.sh
Comment (Huang-Wei) [reply test/cmd/node-management.sh]: Resolved using go-template trick.

**Comment by Huang-Wei** [comment]
@liggitt Thanks for your comments - addressed using a go-template trick. PTAL.

**Comment by liggitt** [comment]
Looks good, thanks

/approve

Will leave lgtm to original reviewers

**Comment by ravisantoshgudimetla** [comment]
/lgtm

Scheduling changes looks good as well.

**Pull request merged**
